src/System.Private.CoreLib/shared/System/Globalization/CharUnicodeInfo.cs

```diff
@@ -71,251 +74,372 @@ internal static int InternalConvertToUtf32(StringBuilder s, int index)
-                        charLength++;
-                        return (temp1 * 0x400) + temp2 + UNICODE_PLANE01_START;
-                    }
-                }
-            }
-            return (int)s[index];
+            nuint offset = GetCategoryCasingTableOffsetNoBoundsChecks(codePoint);
+
+            // Each entry of the 'CategoryValues' table uses bits 5 - 6 to store the strong bidi information.
+
+            StrongBidiCategory bidiCategory = (StrongBidiCategory)(Unsafe.AddByteOffset(ref MemoryMarshal.GetReference(CategoriesValues), offset) & 0b_0110_0000);
+            Debug.Assert(bidiCategory == StrongBidiCategory.Other || bidiCategory == StrongBidiCategory.StrongLeftToRight || bidiCategory == StrongBidiCategory.StrongRightToLeft, "Unknown StrongBidiCategory value.");
+
+            return bidiCategory;
         }
 
-        /// <summary>
-        /// This is called by the public char and string, index versions
-        /// Note that for ch in the range D800-DFFF we just treat it as any
-        /// other non-numeric character
-        /// </summary>
-        internal static double InternalGetNumericValue(int ch)
-        {
-            Debug.Assert(ch >= 0 && ch <= 0x10ffff, "ch is not in valid Unicode range.");
-            // Get the level 2 item from the highest 12 bit (8 - 19) of ch.
-            int index = ch >> 8;
-            if ((uint)index < (uint)NumericLevel1Index.Length)
-            {
-                index = NumericLevel1Index[index];
-                // Get the level 2 offset from the 4 - 7 bit of ch.  This provides the base offset of the level 3 table.
-                // Note that & has the lower precedence than addition, so don't forget the parathesis.
-                index = NumericLevel2Index[(index << 4) + ((ch >> 4) & 0x000f)];
-                index = NumericLevel3Index[(index << 4) + (ch & 0x000f)];
-                ref byte value = ref Unsafe.AsRef(in NumericValues[index * 8]);
-
-                if (BitConverter.IsLittleEndian)
-                {
-                    return Unsafe.ReadUnaligned<double>(ref value);
-                }
+        /*
+         * GetDecimalDigitValue
+         * ====================
+         * Data derived from https://www.unicode.org/reports/tr44/#UnicodeData.txt. If Numeric_Type=Decimal,
+         * then retrieves the Numeric_Value (0..9) for this code point. If Numeric_Type!=Decimal, returns -1.
+         * This data is encoded in field 6 of UnicodeData.txt.
+         */
 
-                return BitConverter.Int64BitsToDouble(BinaryPrimitives.ReverseEndianness(Unsafe.ReadUnaligned<long>(ref value)));
-            }
-            return -1;
+        public static int GetDecimalDigitValue(char ch)
+        {
+            return GetDecimalDigitValueInternalNoBoundsCheck(ch);
         }
 
-        internal static byte InternalGetDigitValues(int ch, int offset)
+        public static int GetDecimalDigitValue(string s, int index)
```

Comment:
>string [](start = 47, length = 6)

should we make this nullable?

Reply:
None of the other APIs allow null parameters - seems like allowing this to be nullable would be an outlier?